File: apps/frontend/src/schemas/scenario.ts
import { z } from "zod";

export const scenarioStatusSchema = z.enum([
  "DRAFT",
  "SIMULATED",
  "SUBMITTED",
  "APPROVED",
  "REJECTED",
  "APPLIED",
]);

export const scenarioConfigSchema = z.object({
  num_robots: z.number().int().min(1).max(10),
  num_tasks: z.number().int().min(1).max(10_000),
  task_arrival_interval: z.number().min(1).max(60),
  travel_time: z.number().positive().max(86_400),
  loading_time: z.number().positive().max(86_400),
  simulation_time: z.number().positive().max(86_400),
  layout_id: z.string().min(1).max(80).default("LAYOUT-DEFAULT"),
  layout_version: z.number().int().min(1).default(3),
  route_id: z.string().min(1).max(80).default("BATTERY_DELIVERY"),
  robot_speed_mps: z.number().positive().max(10).default(1),
  charger_count: z.number().int().min(1).max(20).default(1),
  route_distance_m: z.number().positive().max(100_000).default(30),
  congestion_multiplier: z.number().min(1).max(10).default(1),
});

export const scenarioRunRequestSchema = scenarioConfigSchema.extend({
  name: z.string().trim().min(1, "Name is required").max(80),
});

export const scenarioMetricsSchema = z.object({
  completed_tasks: z.number().int().nonnegative(),
  unfinished_tasks: z.number().int().nonnegative(),
  completion_rate: z.number().min(0).max(1),
  throughput_per_hour: z.number().nonnegative(),
  average_cycle_time: z.number().nonnegative(),
  average_waiting_time: z.number().nonnegative(),
  fleet_utilization_percent: z.number().min(0).max(100),
  starvation_events: z.number().int().nonnegative(),
  congestion_percent: z.number().min(0).max(100),
  travel_distance: z.number().nonnegative(),
  average_delivery_delay: z.number().nonnegative(),
});

const utcDateTimeSchema = z.string().datetime({ offset: true }).refine(
  (value) => value.endsWith("Z"),
  "Expected a UTC timestamp ending in Z",
);

export const scenarioSchema = z.object({
  id: z.string().min(1),
  name: z.string().min(1),
  status: scenarioStatusSchema,
  config: scenarioConfigSchema,
  metrics: scenarioMetricsSchema,
  duration_ms: z.number().nonnegative(),
  created_at: utcDateTimeSchema,
  created_by: z.string().uuid().nullable(),
  reviewed_at: utcDateTimeSchema.nullable(),
  reviewed_by: z.string().uuid().nullable(),
  applied_at: utcDateTimeSchema.nullable(),
  applied_by: z.string().uuid().nullable(),
  version: z.number().int().min(1),
});

export type ScenarioStatus = z.infer<typeof scenarioStatusSchema>;
export type ScenarioConfig = z.infer<typeof scenarioConfigSchema>;
export type ScenarioRunRequest = z.input<typeof scenarioRunRequestSchema>;
export type ScenarioMetrics = z.infer<typeof scenarioMetricsSchema>;
export type Scenario = z.infer<typeof scenarioSchema>;

File: apps/backend/src/ev_twin_api/schemas/scenario.py
from datetime import datetime
from enum import StrEnum
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

from ev_twin_api.schemas.base import UtcDatetime


class ScenarioStatus(StrEnum):
    DRAFT = "DRAFT"
    SIMULATED = "SIMULATED"
    APPROVED = "APPROVED"
    REJECTED = "REJECTED"
    APPLIED = "APPLIED"


class ScenarioConfig(BaseModel):
    """Inputs shared by the SimPy benchmark and the realtime mock twin.

    Only ``num_robots`` and ``task_arrival_interval`` map to the realtime
    MockFactory. Travel/loading/simulation time and task count are benchmark
    inputs only.
    """

    model_config = ConfigDict(str_strip_whitespace=True)

    num_robots: int = Field(ge=1, le=10)
    num_tasks: int = Field(ge=1, le=10_000)
    task_arrival_interval: float = Field(ge=1.0, le=60.0)
    travel_time: float = Field(gt=0.0, le=86_400.0)
    loading_time: float = Field(gt=0.0, le=86_400.0)
    simulation_time: float = Field(gt=0.0, le=86_400.0)
    layout_id: str = Field(default="LAYOUT-DEFAULT", min_length=1, max_length=80)
    layout_version: int = Field(default=1, ge=1)
    route_id: str = Field(default="BATTERY_DELIVERY", min_length=1, max_length=80)
    robot_speed_mps: float = Field(default=1.0, gt=0.0, le=10.0)
    charger_count: int = Field(default=1, ge=1, le=20)
    route_distance_m: float = Field(default=30.0, gt=0.0, le=100_000.0)
    congestion_multiplier: float = Field(default=1.0, ge=1.0, le=10.0)


class ScenarioRunRequest(ScenarioConfig):
    name: str = Field(min_length=1, max_length=80)

    @field_validator("name")
    @classmethod
    def reject_blank_name(cls, value: str) -> str:
        if not value:
            raise ValueError("name must not be blank")
        return value

    def to_config(self) -> ScenarioConfig:
        return ScenarioConfig.model_validate(self.model_dump(exclude={"name"}))


class ScenarioMetrics(BaseModel):
    completed_tasks: int = Field(ge=0)
    unfinished_tasks: int = Field(ge=0)
    completion_rate: float = Field(ge=0.0, le=1.0)
    throughput_per_hour: float = Field(ge=0.0)
    average_cycle_time: float = Field(ge=0.0)
    average_waiting_time: float = Field(ge=0.0)
    fleet_utilization_percent: float = Field(default=0.0, ge=0.0, le=100.0)
    starvation_events: int = Field(default=0, ge=0)
    congestion_percent: float = Field(default=0.0, ge=0.0, le=100.0)
    travel_distance: float = Field(default=0.0, ge=0.0)
    average_delivery_delay: float = Field(default=0.0, ge=0.0)


class Scenario(BaseModel):
    id: str
    name: str
    status: ScenarioStatus
    config: ScenarioConfig
    metrics: ScenarioMetrics
    duration_ms: float = Field(ge=0.0)
    created_at: UtcDatetime
    created_by: UUID | None = None
    reviewed_at: UtcDatetime | None = None
    reviewed_by: UUID | None = None
    applied_at: UtcDatetime | None = None
    applied_by: UUID | None = None
    version: int = Field(default=1, ge=1)

    def with_status(
        self,
        status: ScenarioStatus,
        *,
        reviewed_at: datetime | None = None,
        reviewed_by: UUID | None = None,
        applied_at: datetime | None = None,
        applied_by: UUID | None = None,
    ) -> "Scenario":
        updates: dict[str, object] = {
            "status": status,
            "version": self.version + 1,
        }
        if reviewed_at is not None:
            updates["reviewed_at"] = reviewed_at
        if reviewed_by is not None:
            updates["reviewed_by"] = reviewed_by
        if applied_at is not None:
            updates["applied_at"] = applied_at
        if applied_by is not None:
            updates["applied_by"] = applied_by
        return self.model_copy(update=updates, deep=True)
